Migrate components/file_preview to typescript

compiledsound opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Small changes to enable typescript in the file preview component.
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-20509
Fixes https://github.com/mattermost/mattermost-server/issues/13730

mattermod commented:
Thanks @compiledsound for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

devinbinnie commented on components/file_preview/file_preview.test.tsx:
```diff
@@ -69,6 +71,9 @@ describe('FilePreview', () => {
         );
+
+        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+        //@ts-ignore
```
This is a big code smell, what's the error being thrown here?
If it's around the `shallow()` method, you can use the generic type `ShallowWrapper` to define your component.

compiledsound replied:
the issue is that the handleRemove method does not exist on wrapper.instance(). I found a similar issue from 2018 and one suggestion from the post is to cast the instance.

devinbinnie commented on components/file_preview/file_preview.tsx:
```diff
@@ -1,38 +1,32 @@
+    enableSVGs: boolean;
+    onRemove: (id: string) => void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fileInfos: any[];
```
Instead of using an `any` here, we can define a `FilePreviewInfo` type in this file for use by this file. It should contain whatever fields are currently consumed by the component.

compiledsound replied:
is there a list of fields defined somewhere? when I was looking it seemed to vary so I wasn't sure.

devinbinnie replied:
You can define whatever the `file_preview.tsx` component consumes, and set the fields that are optional to nullable.

devinbinnie commented on components/file_preview/file_preview.tsx:
```diff
@@ -115,7 +109,7 @@ export default class FilePreview extends React.PureComponent {
                     key={clientId}
                     clientId={clientId}
-                    fileInfo={this.props.uploadsProgressPercent[clientId]}
+                    fileInfo={this.props.uploadsProgressPercent.get(clientId)}
```
We shouldn't be changing the logic of how the component works with this migration.
I believe you can use the type `Record<string, FilePreviewInfo>`

devinbinnie commented on components/file_preview/file_progress_preview.tsx:
```diff
@@ -11,13 +9,14 @@ import RemoveIcon from 'components/widgets/icons/fa_remove_icon';
+    handleRemove: (id: string) => void;
+    clientId: string;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fileInfo: any;
```
Also of type `FilePreviewInfo`

devinbinnie commented on package-lock.json:
```diff
@@ -5657,7 +5657,7 @@
         },
         "p-cancelable": {
           "version": "0.4.1",
-          "resolved": "http://registry.npmjs.org/p-cancelable/-/p-cancelable-0.4.1.tgz",
```
These `package-lock` changes aren't required, can we revert them?

compiledsound replied:
will do, seemed like this was an automatic change by vs code

devinbinnie commented on components/file_preview/file_preview.tsx:
```diff
@@ -1,38 +1,32 @@
+    fileInfos: any[];
+    uploadsInProgress: string[];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    uploadsProgressPercent: Map<string, any>;
```
Same with this type, which should also be `FilePreviewInfo`

mickmister replied:
These two props were optional, and now are required. Is this intentional?

Also we've changed the Props contract on the datatype here. Is the parent supposed to pass in an object or a Map? For an object you can use this type:

```ts
uploadsProgressPercent: {[clientID: string]: FilePreviewInfo}
```

devinbinnie commented on components/file_preview/file_preview.test.tsx:
```diff
@@ -20,13 +20,15 @@ describe('FilePreview', () => {
-            extension: 'image/png',
-        },
-    };
+    const uploadsProgressPercent = new Map(Object.entries(
```
This shouldn't need to change, should remain a POJO.

lindy65 commented:
Thanks @compiledsound @devinbinnie - I'll QA once dev review is complete. Please mention me here when ready for QA. Thanks!

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4898.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mickmister commented on components/file_preview/file_preview.tsx:
```diff
@@ -1,38 +1,32 @@
-    static defaultProps = {
-        fileInfos: [],
-        uploadsInProgress: [],
-        uploadsProgressPercent: {},
```
I think we should keep the `defaultProps`. The parent component is not matched up to respect this component's props yet, so it may not be passing in these props properly.

mickmister replied:
I think these are supposed to be optional anyway

compiledsound replied:
restored defaultProps and made properties optional

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `3dbc70fe359ef57dc5443de6f8cadaaa6df171c5`.

Access here: https://mattermost-webapp-pr-4898.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `d9a64b42e490b13984f4a339f970dd7860a90ac1`.

Access here: https://mattermost-webapp-pr-4898.test.mattermost.cloud

devinbinnie commented on components/file_preview/file_preview.tsx:
```diff
@@ -110,16 +121,21 @@ export default class FilePreview extends React.PureComponent {
-        });
+        if (this.props.uploadsInProgress) {
+            this.props.uploadsInProgress.forEach((clientId) => {
+                if (this.props.uploadsProgressPercent) {
```
Since this doesn't depend on an iteration of the loop, you can null-check this outside of the loop.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

compiledsound commented:
@devinbinnie would you know what the solution to the build failure is? I don't see how my changes could have caused it.

devinbinnie commented:
> @devinbinnie would you know what the solution to the build failure is? I don't see how my changes could have caused it.

There's a bug in `mattermost-redux` which will need to be fixed before the install phase will finished. Sorry about that, nothing you did.

mattermod commented:
Test server destroyed

devinbinnie commented:
/update-branch

devinbinnie commented:
@compiledsound I've updated the branch for you which should fix the redux issue.

compiledsound commented:
@devinbinnie this is a new workflow for me.... how can I sync with master? I fixed some formatting errors and thought I rebased onto an updated master, but it seems I did not in my latest push.

devinbinnie commented:
@compiledsound I updated the branch for you, so all you needed to do was pull in my changes. I can do it again for you, this time, just run a `git pull` before you do anything to the repo.

devinbinnie commented:
/update-branch

compiledsound commented:
@lindy65 ready for QA

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4898.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Test server destroyed
